### PR TITLE
add example diff command

### PR DESCRIPTION
**Pull request by twpayne**
This command serves as an usage example and a quick way to manually test changes to the library.

### CODE REVIEW

File: cmd/diff/main.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
+package main
```
Comment (josharian) [inline cmd/diff/main.go]: This needs package docs.

File: cmd/diff/main.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	"flag"
+	"fmt"
+	"os"
+
+	"github.com/pkg/diff"
+)
+
+var (
+	color       = flag.Bool("color", false, "color")
+	contextSize = flag.Int("context", 3, "context size")
```
Comment (josharian) [inline cmd/diff/main.go]: What do you think about `-u` instead of `-context`, to match traditional `diff`?

File: cmd/diff/main.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+)
+
+func fileLines(filename string) ([]string, error) {
```
Comment (josharian) [inline cmd/diff/main.go]: short doc please

File: cmd/diff/main.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	}
+	if *color {
+		opts = append(opts, diff.TerminalColor())
+	}
+	_, err = e.WriteUnified(os.Stdout, ab, opts...)
+	return err
+}
+
+func main() {
+	if err := run(); err != nil {
```
Comment (josharian) [inline cmd/diff/main.go]: I don't think this wrapper buys us much. I'd suggest instead something like:

```func check(err error) {
  if err != nil {
    log.Fatal(err)
  }
}
```

and then turning `func run` into `func main`, and calling `check(err)` as needed.

File: cmd/diff/main.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	}
+	return lines, s.Err()
+}
+
+func run() error {
+	ctx := context.Background()
+
+	flag.Parse()
+	if len(flag.Args()) != 2 {
+		return fmt.Errorf("syntax: %s filename1 filename2", os.Args[0])
```
Comment (josharian) [inline cmd/diff/main.go]: This should be something more like:

```go
fmt.Fprintln(os.Stderr, "diff filename1 filename2")
os.Exit(2)
```

File: cmd/diff/main.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	var lines []string
+	s := bufio.NewScanner(f)
+	for s.Scan() {
+		lines = append(lines, s.Text())
+	}
+	return lines, s.Err()
+}
+
+func run() error {
+	ctx := context.Background()
```
Comment (josharian) [inline cmd/diff/main.go]: Let's either inline this in the `diff.Myers` call or place it in the line or two immediately preceding it. (And possibly add a timeout flag and construct a deadline context from it?)

File: cmd/diff/main.go (hunk 1 above)
Comment (twpayne) [reply cmd/diff/main.go]: Done.

File: cmd/diff/main.go (hunk 2 above)
Comment (twpayne) [reply cmd/diff/main.go]: Done.

File: cmd/diff/main.go (hunk 3 above)
Comment (twpayne) [reply cmd/diff/main.go]: Done.

File: cmd/diff/main.go (hunk 4 above)
Comment (twpayne) [reply cmd/diff/main.go]: Done.

File: cmd/diff/main.go (hunk 5 above)
Comment (twpayne) [reply cmd/diff/main.go]: Done.

File: cmd/diff/main.go (hunk 6 above)
Comment (twpayne) [reply cmd/diff/main.go]: Done.

**Comment by twpayne** [comment]
Thanks for the feedback, I've updated the PR.

I suggest that we keep this command as a quick-and-dirty way of testing `github.com/pkg/diff`, and explicitly avoid trying to match the functionality of the `diff` command exactly. Specifically, Go's standard `flag` library is not rich enough to handle all of the options to the `diff` command (where there are multiple options with multiple short versions and optional arguments).

**Comment by josharian** [comment]
Thanks! Very much appreciated. I have a few minor remaining comments, but I'll just make some tweaks in a follow-up commit.

**Comment by josharian** [comment]
Out of curiosity, how did you come across this package and what are you using it for? (Feel free not to answer either question.)

**Pull request merged**